### PR TITLE
Vertex shader debugger

### PR DESCRIPTION
**Pull request by neobrain**
Adds a simple shader disassembly widget. Some fine-tuning would still be useful probably, but I wanted to get this out so I can focus on other things for now.

Screenshot:
Pica Shader Debugger

Pretty sure Windows build is broken once again, so some error messages or build fix patches would be appreciated :)

### CODE REVIEW

File: src/citra_qt/debugger/graphics_vertex_shader.cpp [hunk 1]
```diff
@@ -0,0 +1,298 @@
+// Copyright 2014 Citra Emulator Project
+// Licensed under GPLv2
```
Comment (turbedi) [inline src/citra_qt/debugger/graphics_vertex_shader.cpp]: or any later version(?)

File: src/citra_qt/debugger/graphics_vertex_shader.hxx [hunk 2]
```diff
@@ -0,0 +1,51 @@
+// Copyright 2014 Citra Emulator Project
+// Licensed under GPLv2
```
Comment (turbedi) [inline src/citra_qt/debugger/graphics_vertex_shader.hxx]: or any later version(?)

File: src/citra_qt/debugger/graphics_vertex_shader.cpp (hunk 1 above)
Comment (neobrain) [inline src/citra_qt/debugger/graphics_vertex_shader.cpp]: Thanks, fixed:)

File: src/citra_qt/debugger/graphics_vertex_shader.cpp [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+                                                 const std::string& address_register_name) {
+                std::string relative_address;
+                if (!address_register_name.empty())
+                    relative_address = "[" + address_register_name + "]";
+
+                output << std::setw(10) << std::right << (negate ? "-" : "") + input.GetName() + relative_address;
+                output << "." << swizzle_mask;
+            };
+
+            static auto print_input_indexed_compact = [](std::stringstream& output, const SourceRegister& input,
```
Comment (bunnei) [inline src/citra_qt/debugger/graphics_vertex_shader.cpp]: Why do the others have examples but not this one?

File: src/citra_qt/debugger/graphics_vertex_shader.cpp [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+        }
+
+        break;
+    }
+    }
+
+    return QVariant();
+}
+
+QVariant GraphicsVertexShaderModel::data(const QModelIndex& index, int role) const {
```
Comment (bunnei) [inline src/citra_qt/debugger/graphics_vertex_shader.cpp]: There seems to be a lot of code in this function for pretty-printing the disassembled output. Wouldn't this be helpful in nhistro?

File: src/citra_qt/debugger/graphics_vertex_shader.cpp [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+                    if (instr.opcode.GetInfo().subtype & Instruction::OpCodeInfo::HasAlternative) {
+                        output << "else jump to 0x" << std::setw(4) << std::right << std::setfill('0') << 4 * instr.flow_control.dest_offset << " ";
+                    } else if (instr.opcode.GetInfo().subtype & Instruction::OpCodeInfo::HasExplicitDest) {
+                        output << "jump to 0x" << std::setw(4) << std::right << std::setfill('0') << 4 * instr.flow_control.dest_offset << " ";
+                    } else {
+                        // TODO: Handle other cases
+                    }
+
+                    if (instr.opcode.GetInfo().subtype & Instruction::OpCodeInfo::HasFinishPoint) {
+                        output << "(return on " << std::setw(4) << std::right << std::setfill('0') << 4 * instr.flow_control.dest_offset + 4 * instr.flow_control.num_instructions << ")";
```
Comment (bunnei) [inline src/citra_qt/debugger/graphics_vertex_shader.cpp]: Huge line :|

**Comment by bunnei** [comment]
Tested on Windows. Works fine and looks mostly fine. A couple of nitpicks:
* The text in the vertex shader debugger is not aligned on Windows (unless you set some global monospaced font I guess?) Looks aligned on the CPU disasm though, so should be easy enough to fix. (Although seriously, why don't we just have a monospaced font for the disasm widgets?)
* Alternate the lines light gray/white - This makes it slightly more readable, but more importantly IMO makes it stylistically consistent with the other GUI widgets

File: src/citra_qt/debugger/graphics_vertex_shader.cpp (hunk 4 above)
Comment (neobrain) [inline src/citra_qt/debugger/graphics_vertex_shader.cpp]: In theory maybe, but in practice this restricts us from translating the shader debugger. I also consider the actual representation of data to be orthogonal to nihstro's core purpose.

**Comment by neobrain** [comment]
Ready.

**Comment by neobrain** [comment]
(the monospace font issue needs to be addresses separately btw, since it affects virtually all widgets)

**Pull request merged**